_posts/2022-01-31-how-we-ship-ios-libraries.md

```diff
@@ -0,0 +1,333 @@
+
+The premise of our application is easy:
+
+> Show a non-interactive message to the user with data from a Rust library.
+
+Let's get started on that.
+
+## The project
+
+We start with a fresh iOS project.
+Go to File -> New -> Project, then choose the iOS App template,
+give it a name such as `ShippingRust`,
+select where to store it and finally create it.
+You're greeted with `ContentView.swift` and the following code:
+
+```swift
+import SwiftUI
+
+struct ContentView: View {
+    var body: some View {
+        Text("Hello, world!")
+            .padding()
+    }
+}
+```
+
+You can build and run it now. This will open the Simulator and display "Hello, world!".
+We'll get back to the Swift application later.
+
+## The Rust parts
+
+First we set up the Rust library.
+
+In a terminal navigate to your `ShippingRust` project directory.
+In there create a new Rust crate:
+
+```
+cargo new --lib shipping-rust-ffi
+```
+
+We will need a static library, so we change the crate type in the generated `shipping-rust-ffi/Cargo.toml`.
+Add the following below the package configuration:
+
+```
+[lib]
+crate-type = ["staticlib"]
+```
+
+Let's also turn the project into a Cargo workspace.
+Create a new top-level `Cargo.toml` with the content:
+
+```
+[workspace]
+members = [
+  "shipping-rust-ffi"
+]
+```
+
+`cargo build` in the project directory should work now and create a new static library.
+
+```
+; ls -l target/debug/libshipping_rust_ffi.a
+-rw-r--r-- 2 jer staff 16061952 Jan 28 13:09 target/debug/libshipping_rust_ffi.a
+```
+
+Let's add some code to `shipping-rust-ffi/src/lib.rs` next.
+Nothing fancy, a simple function taking some arguments and returning the sum:
+
+```rust
+use std::os::raw::c_int;
+
+#[no_mangle]
+pub extern "C" fn shipping_rust_addition(a: c_int, b: c_int) -> c_int {
+    a + b
+}
+```
+
+The `no_mangle` ensures the name lands in the compiled library as is
+and the `extern "C"` makes sure it uses the right ABI.
+
+We now got a Rust library, exporting a C-ABI compatible interface.
+We can now consume this in our iOS application.
+
+## The Xcode parts
+
+Before we can use the code we need a bit more setup.
+
+We start by linking against the `libshipping_rust_ffi.a` library.
+In your Xcode project open your target configuration[^1],
+go to "Build Phases", then look for "Link Binary with Libraries".
+Add a new one, in the popup select "Add files" on the bottom left
+and look for the `target/debug/libshipping_rust_ffi.a` file.
+Yes, that's actually for the wrong target. This is just for the name, we'll fix up the path next.
+Go to "Build Settings" and search for "Library Search Paths".
+It probably has the path to file in there right now for both `Debug` and `Release` builds.
+Remove that one for `Debug`, then add a new row by clicking the small `+` symbol.
+Select any matcher, e.g. `Any Driverkit`. It doesn't matter what value you give it.
+We will manually modify that entry next.
+Do the same for the `Release` configuration.
```

Comment:
There are so many fiddly user-interface-dependent steps! :(

Reply:
Welcome to the Xcode world, where half the things are done in the GUI or you can try to fiddle with the pbxproj file manually, where the format has tiny differences between Xcode versions.